pkg/occlient/occlient.go

```diff
@@ -2006,67 +2009,103 @@ func (c *Client) DeleteServiceInstance(labels map[string]string) error {
 	if err != nil {
-		return errors.Wrapf(err, "unable to watch project")
+		return errors.Wrap(err, "unable to delete project")
 	}
 
-	defer w.Stop()
-	for {
-		val, ok := <-w.ResultChan()
-		// When marked for deletion... val looks like:
-		/*
-			val: {
-				Type:MODIFIED
-				Object:&Project{
-					ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{...},
-					Spec:ProjectSpec{...},
-					Status:ProjectStatus{
-						Phase:Terminating,
-					},
-				}
-			}
-		*/
-		// Post deletion val will look like:
-		/*
-			val: {
-				Type:DELETED
-				Object:&Project{
-					ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{...},
-					Spec:ProjectSpec{...},
-					Status:ProjectStatus{
-						Phase:,
-					},
-				}
-			}
-		*/
-		if !ok {
-			return fmt.Errorf("received unexpected signal %+v on project watch channel", val)
-		}
-		// So we depend on val.Type as val.Object.Status.Phase is just empty string and not a mapped value constant
-		if prj, ok := val.Object.(*projectv1.Project); ok {
-			glog.V(4).Infof("Status of delete of project %s is %s", name, prj.Status.Phase)
-			switch prj.Status.Phase {
-			//prj.Status.Phase can only be "Terminating" or "Active" or ""
-			case "":
-				if val.Type == watch.Deleted {
-					return nil
+	// If watcher has been created (wait was passed) we will create a go routine and actually **wait**
+	// until *EVERYTHING* is successfully deleted.
+	if watcher != nil {
+
+		// Project channel
+		// Watch error channel
+		projectChannel := make(chan *projectv1.Project)
+		watchErrorChannel := make(chan error)
+
+		// Create a go routine to run in the background
+		go func() {
+
+		loop:
+			for {
+
+				// If watch unexpected has been closed..
+				val, ok := <-watcher.ResultChan()
+				if !ok {
+					//return fmt.Errorf("received unexpected signal %+v on project watch channel", val)
+					watchErrorChannel <- errors.Errorf("watch channel was closed unexpectedly: %+v", val)
+					break loop
 				}
-				if val.Type == watch.Error {
-					return fmt.Errorf("failed watching the deletion of project %s", name)
+
+				// So we depend on val.Type as val.Object.Status.Phase is just empty string and not a mapped value constant
+				if projectStatus, ok := val.Object.(*projectv1.Project); ok {
+
+					glog.V(4).Infof("Status of delete of project %s is '%s'", name, projectStatus.Status.Phase)
+
+					switch projectStatus.Status.Phase {
+					//projectStatus.Status.Phase can only be "Terminating" or "Active" or ""
+					case "":
+						if val.Type == watch.Deleted {
+							projectChannel <- projectStatus
+							break loop
+						}
+						if val.Type == watch.Error {
+							watchErrorChannel <- errors.Errorf("failed watching the deletion of project %s", name)
+							break loop
+						}
+					}
+
+				} else {
+					watchErrorChannel <- errors.New("unable to convert event object to Project")
+					break loop
 				}
+
 			}
+			close(projectChannel)
```

Comment:
Maybe close it right after creation like `defer close(projectChannel)`?

Reply:
IMO, I like keeping the close statements at the end instead of the beginning (especially since this just for the go routine).

Reply:
https://groups.google.com/forum/#!topic/golang-nuts/Hj7-HV-W_iU
I went through the above discussion. If a panic happens in the go routine, we wouldn't reach the close statements and thus the channels won't close.

Reply:
Okay, this is how we currently have it for the PodTimeout go routine that's been written. Should we change this in both locations?

Reply:
I would suggest a different PR but I leave the decision up to you

Reply:
> Maybe close it right after creation like `defer close(projectChannel)`?

+1

Reply:
@cdrage Can you please create a issue regarding this discussion?